File: android/core/src/main/kotlin/com/tickernest/core/Money.kt
package com.tickernest.core

import java.math.BigDecimal
import java.math.MathContext
import java.math.RoundingMode

/**
 * Money is BigDecimal everywhere — never Double / Float. Mirrors the
 * NUMERIC(20,4) backend invariant.
 *
 * Construct via D() so rounding mode is centralised; all financial maths
 * goes through the helpers here.
 */
private val CTX = MathContext(30, RoundingMode.HALF_EVEN)

typealias Money = BigDecimal

@Suppress("FunctionName")
fun D(v: String): Money = BigDecimal(v, CTX)
@Suppress("FunctionName")
fun D(v: Int): Money = BigDecimal(v).round(CTX)
@Suppress("FunctionName")
fun D(v: Long): Money = BigDecimal(v).round(CTX)
@Suppress("FunctionName")
fun D(v: Double): Money = BigDecimal(v.toString(), CTX) // string ctor avoids float-binary surprises

val ZERO: Money = BigDecimal.ZERO

fun Money.isZero(): Boolean = this.signum() == 0

fun Iterable<Money>.sumMoney(): Money =
    fold(ZERO) { a, b -> a.add(b) }

/**
 * Weighted average: sum(qty_i * price_i) / sum(qty_i).
 * Returns ZERO when totalQty is 0 — never NaN, never throws.
 */
fun weightedAvg(pairs: List<Pair<Money, Money>>): Money {
    val totalQty = pairs.map { it.first }.sumMoney()
    if (totalQty.isZero()) return ZERO
    val numerator = pairs.map { it.first.multiply(it.second, CTX) }.sumMoney()
    return numerator.divide(totalQty, CTX)
}

fun Money.toWire(): String = setScale(4, RoundingMode.HALF_EVEN).toPlainString()

File: services/api/src/common/types/money.ts
import { Decimal } from 'decimal.js';

/**
 * Money is NEVER a number in this codebase. Always Decimal or string.
 * DB type: NUMERIC(20,4). Wire type: string.
 *
 * Construct via D() to centralise rounding mode; all financial maths runs
 * through these helpers so we can audit precision in one place.
 */
Decimal.set({ precision: 30, rounding: Decimal.ROUND_HALF_EVEN });

export type Money = Decimal;
export const D = (v: Decimal.Value): Money => new Decimal(v ?? 0);
export const ZERO: Money = D(0);

export const isZero = (m: Money): boolean => m.isZero();
export const sum = (xs: Money[]): Money => xs.reduce((a, b) => a.add(b), ZERO);

/**
 * Weighted average: sum(qty_i * price_i) / sum(qty_i).
 * Returns ZERO when totalQty is 0 — never NaN, never throws.
 */
export const weightedAvg = (
  pairs: ReadonlyArray<{ qty: Money; price: Money }>,
): Money => {
  const totalQty = sum(pairs.map((p) => p.qty));
  if (totalQty.isZero()) return ZERO;
  const numerator = sum(pairs.map((p) => p.qty.mul(p.price)));
  return numerator.div(totalQty);
};

export const toWire = (m: Money): string => m.toFixed(4);
